modules/compiler/vecz/source/analysis/uniform_value_analysis.cpp

```diff
@@ -76,6 +77,32 @@ bool isDivergenceReduction(const Function &F) {
           L.Consume("divergence_"));
 }
 
+bool isTrueUniformInternal(const Value *V, unsigned Depth) {
+  // Arguments and Constants are always uniform
+  if (isa<Argument>(V) || isa<Constant>(V)) {
+    return true;
+  }
+
+  constexpr unsigned DepthLimit = 6;
+
+  if (Depth < DepthLimit) {
+    // For a specific subset of instructions, if all operands are truly
+    // uniform, then the instruction is too.
+    // FIXME: This is pessimistic.
+    if (const auto *I = dyn_cast<Instruction>(V)) {
+      if (isa<UnaryOperator>(I) || isa<BinaryOperator>(I) || isa<CastInst>(I) ||
+          isa<CmpInst>(I) || isa<FreezeInst>(I) || isa<SelectInst>(I) ||
+          isa<PHINode>(I)) {
```

Comment:
This looks suspicious. `freeze` is used to handle `undef` and `poison`, but if we take `undef` and `poison` into account, is the logic still correct for any instructions? `freeze undef` and `freeze undef` are not guaranteed to be equal.

Reply:
Hmm you're probably right. I was initially thinking that each instance of `freeze` would be uniform across all _lanes_, even if it wasn't the exact same value between individual instances of the `freeze` instruction. On second though, I'm not sure this holds: once vectorized, `freeze <4 x i32> undef` isn't guaranteed to return the same value in all "lanes". It probably does, but we can't rely on it.

I'm not 100% certain that this holds. The problem is that the slow SYCL-DNN benchmark starts off with a `freeze` of an incoming argument. Maybe we can special case `freeze` to handle certain specific cases, I dunno. If not, then we're in a bit of trouble.

What do you think?

Reply:
I think the logic holds if we don't have `undef` operands and the instructions do not produce `poison` or `undef` either. Clang will often add `noundef` to incoming arguments. Maybe checking that will be good enough?

Reply:
Yeah that's probably the way to go. 

Since we only check arguments and constants as the "leaves" we're quite easily able to prevent `undef`, `poison` or arguments without `noundef`.

Of course, we do still have to ensure that we don't end up seeing a `udiv 1, 0` or whatever. We shouldn't rely on expecting LLVM to have already replaced those with `poison`. I know there are ways of getting LLVM to tell us whether something is guaranteed not to produce poison/undef but I'll have to see if they're readily available.

Reply:
I think I've addressed this now @hvdijk. No more handling of freeze, and checking that everything we judge to be truly uniform is also guaranteed not to be undef or poison.

Reply:
Looks good. We might also say, we do allow `freeze` instructions but only if the *operand* (as opposed to the result) is known not to be `undef` or `poison`. But not sure we have any test that would benefit from it.

Reply:
Yes, that could also work. It would be interesting to find a test case for that. Most likely LLVM won't add a `freeze` to something it already knows is not poison/undef (at least from what I've seen in the sycl-dnn test I'm fixing with this) but it could happen through other means.